program/src/processor.rs

```diff
@@ -2311,6 +2313,385 @@ impl Processor {
+                spot_market_ai.clone(),
+                market_open_orders_ai.clone(),
+                signer_ai.clone(),
+                dex_base_ai.clone(),
+                dex_quote_ai.clone(),
+                base_root_bank_ai.clone(),
+                base_node_bank_ai.clone(),
+                quote_root_bank_ai.clone(),
+                quote_node_bank_ai.clone(),
+                base_vault_ai.clone(),
+                quote_vault_ai.clone(),
+                dex_signer_ai.clone(),
+                token_prog_ai.clone(),
+            ];
+            Self::settle_funds(program_id, &settle_accounts[..])?;
+        }
+
+        {
+            Self::place_spot_order2(program_id, accounts, new_order)?;
+        }
+
+        Ok(())
+    }
+
+    #[inline(never)]
+    pub fn edit_perp_order_by_client_id(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        price: i64,
+        mut max_base_quantity: i64,
+        max_quote_quantity: i64,
+        expiry_timestamp: u64,
+        client_order_id: u64,
+        expected_cancel_size: i64,
+        side: Side,
+        order_type: OrderType,
+        reduce_only: bool,
+        limit: u8,
+    ) -> MangoResult {
+        const NUM_FIXED: usize = 9;
+        let (fixed_ais, _packed_open_orders_ais) = array_refs![accounts, NUM_FIXED; ..;];
+        let [
+            mango_group_ai,             // read
+            mango_account_ai,           // write
+            owner_ai,                   // read, signer
+            _mango_cache_ai,             // read
+            perp_market_ai,             // write
+            bids_ai,                    // write
+            asks_ai,                    // write
+            _event_queue_ai,             // write
+            _referrer_mango_account_ai,  // write
+        ] = fixed_ais;
+
+        {
+            let mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+
+            let mango_account =
+                MangoAccount::load_checked(mango_account_ai, program_id, mango_group_ai.key)?;
+            check!(!mango_account.is_bankrupt, MangoErrorCode::Bankrupt)?;
+            check!(owner_ai.is_signer, MangoErrorCode::SignerNecessary)?;
+            check!(
+                &mango_account.owner == owner_ai.key || &mango_account.delegate == owner_ai.key,
+                MangoErrorCode::InvalidOwner
+            )?;
+
+            let perp_market =
+                PerpMarket::load_checked(perp_market_ai, program_id, mango_group_ai.key)?;
+
+            let market_index = mango_group.find_perp_market_index(perp_market_ai.key).unwrap();
+
+            let book = Book::load_checked(program_id, bids_ai, asks_ai, &perp_market)?;
+
+            let (order_id, cancel_side) = mango_account
+                .find_order_with_client_id(market_index, client_order_id)
+                .ok_or(throw_err!(MangoErrorCode::ClientIdNotFound))?;
+
+            let book_order = book.get_order(order_id, side)?;
```

Comment:
this needs to deal with expired orders: maybe the old order is still on the book, but expired?

Reply:
I thought it's ok to edit/update an expired order as long as the new order is not already expired (and ```place_perp_order2``` checks for this)? But let me know if this is not the case. I'm not sure what the behavior should be if the order is expired (maybe just throw an error?)